Docs: Added Developer FAQ section with first question. (#1202)
bareos/bareos: docs/manuals/source/DeveloperGuide/developerFaq.rst

.. _developer-guide-faq:

Developer FAQ
=============

How to run a job that automatically fails?
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
I want to run a job that automatically fails to simulate failing jobs.

To do so, you can use the directive :config:option:`dir/job/ClientRunBeforeJob`.
This directive launches a script on the client-side ( file daemon ) before running the job.
set the script to be “/bin/false”

.. code-block:: bareosconfig
   :caption: Job with client run before job directive

   Job {
       Name = "backup-bareos-fd"
       JobDefs = "DefaultJob"
       Client = "bareos-fd"
       Client Run Before Job = /bin/false
   }

or, if you want a job to work properly but fail at the end (e.g: simulate a full backup that did not go well), you can do the following

.. code-block:: bareosconfig
   :caption: Job with failing runscript at the end

   Job {
     Name = "backup"
     JobDefs = DefaultJob
     RunScript {
       Command = "/bin/false"
       failjobonerror = yes
       RunsWhen = After
     }
    }

> 

Reviewer: I understand what you mean, but I meant to do them differently to show there are different ways to run scripts, whether with `RunScript` or `Client Run Before/After Job`.

Unless you feel one way should be avoided maybe?

I have changes that affect tapes and autochangers, how can I test them?
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> In order to test tapes, we use mhvtl to simulate the presence of tape and autochanger devices.

Reviewer: it's not.

Documentation is available at https://www.mhvtl.com/

> First you need to clone the mhvtl repository from https://github.com/markh794/mhvtl.git.

Reviewer 2: ```suggestion
Some distribution offer mhvtl package like openSUSE, you can then use your package manager to install and use it.

Otherwise you need to clone the `mhvtl repository <https://github.com/markh794/mhvtl.git>`.
```

Then you should compile and install the mhvtl kernel module

.. code-block:: shell-session

>    $ cd mhvtl/kernel

Reviewer 2: ```suggestion
   $ git clone https://github.com/markh794/mhvtl.git
   $ cd mhvtl/kernel
```
Add clone step.

   $ make
   $ sudo make install

After that you have to install the mhvtl user tools

.. code-block:: shell-session

   $ cd mhvtl
   $ make
   $ sudo make install

Now that mhvtl is installed, you can enable and start the mhvtl services

.. code-block:: shell-session

>    $ sudo systemctl enable mhvtl-load-modules.service
>    $ sudo systemctl start mhvtl-load-modules.service
>    $ sudo systemctl enable mhvtl.target
>    $ sudo systemctl start mhvtl.target

Reviewer 2: The syntax to enable and start a service can be shortened like this with the `--now` flag
```suggestion
   $ sudo systemctl enable --now mhvtl-load-modules.service
   $ sudo systemctl enable --now mhvtl.target
```

> In order to check if your installation was done correctly, you can run :command:`lsscsi` (install it if you don't have it) or :command:`ls` in :file:`/dev/tape/by-id` which should give you somehting similar to the following output

Reviewer 2: ```suggestion
In order to check if your installation was done correctly, you can run :command:`lsscsi` (install it if you don't have it) or :command:`ls /dev/tape/by-id` which should give you something similar to the following output
```

   [2:0:4:0]    tape    IBM      ULT3580-TD8      0107  /dev/st6
   [2:0:8:0]    mediumx STK      L80              0107  /dev/sch0
   [2:0:9:0]    tape    STK      T10000B          0107  /dev/st7
   [2:0:10:0]   tape    STK      T10000B          0107  /dev/st4
   [2:0:11:0]   tape    STK      T10000B          0107  /dev/st3
   [2:0:12:0]   tape    STK      T10000B          0107  /dev/st5

   $ ls /dev/tape/by-id/
   scsi-350223344ab000100      scsi-350223344ab001000      scsi-XYZZY_A3
   scsi-350223344ab000100-nst  scsi-350223344ab001000-nst  scsi-XYZZY_A3-nst
   scsi-350223344ab000200      scsi-350223344ab001100      scsi-XYZZY_A4
   scsi-350223344ab000200-nst  scsi-350223344ab001100-nst  scsi-XYZZY_A4-nst
   scsi-350223344ab000300      scsi-SSTK_L700_XYZZY_A      scsi-XYZZY_B1
   scsi-350223344ab000300-nst  scsi-SSTK_L80_XYZZY_B       scsi-XYZZY_B1-nst
   scsi-350223344ab000400      scsi-XYZZY_A1               scsi-XYZZY_B2
   scsi-350223344ab000400-nst  scsi-XYZZY_A1-nst           scsi-XYZZY_B2-nst
   scsi-350223344ab000900      scsi-XYZZY_A2               scsi-XYZZY_B3
   scsi-350223344ab000900-nst  scsi-XYZZY_A2-nst           scsi-XYZZY_B3-nst

> Now that the installtion is done, you will have to let bareos know that you are building tape and autochanger tests.
> To do so, you'll have to add the needed devices as cmake parameters to your exisiting cmake configuration:

Reviewer 2: ```suggestion
Now that the installation is done, you will have to let bareos know that you are building tape and autochanger tests by adding the needed devices as cmake parameters to your existing cmake configuration:
```

.. code-block:: shell-session

   -Dchanger-device=/dev/tape/by-id/scsi-SSTK_L700_XYZZY_A
   -Dtape-devices=/dev/tape/by-id/scsi-350223344ab000100-nst;/dev/tape/by-id/scsi-350223344ab000200-nst;/dev/tape/by-id/scsi-350223344ab000300-nst;/dev/tape/by-id/scsi-350223344ab000400-nst

The selected devices are the ones used in our tests.

> In case of missing devices or other incoherences, a reboot of the machine can solve most of your issues.

Reviewer 2: ```suggestion
A machine reboot can solve most of issues like missing devices or other incoherences.
```

> In case you update your kernel and kernel headers, you will have to rebuild and reinstall mhvtl, otherwise mhvtl will not work properly and bareos would not build.

Reviewer 2: Maybe we can use stronger words.
```suggestion
For **any update** of kernel and kernel headers, you will have to rebuild and reinstall mhvtl, otherwise mhvtl kernel module loading will fail and command:`mhvtl` will not work properly and bareos will not build.
```

> In certain rare cases, after an mhvtl update (pulling new changes and building), certain tape names change and cause tests to fail. You should keep an eye on those. If such a thing happens, you can check :file:`/etc/mhvtl/device.conf` and modify the device names when necessary.

Reviewer 2: Maybe a shorten sentence?
```suggestion
command:`mhvtl`updates in certain rare cases cause tape names changes causing test to fail. In that case, check :file:`/etc/mhvtl/device.conf` and modify the device names according.
```

.. _developer-guide-faq:

Developer FAQ
=============

Job run
-------

How to run a job that automatically fails?
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> I want to run a job that automatically fails in order to test rescheduling on error.

Reviewer: ```suggestion
I want to run a job that automatically fails for testing purposes.
```
The question is too specific, keeping only the failing is enough in my opinion. The scheduling can be easily found using the documentation. The only new trick here is the `/bin/false`

To run a job that automatically fails, you need to use the directive :config:option:`dir/job/ClientRunBeforeJob`.
This directive launches a script on the client-side ( file daemon ) before running the job.
set the script to be “/bin/false”

.. code-block:: sh
   :caption: Job with client run before job directive

    Job {
        Name = "backup-bareos-fd"
        JobDefs = "DefaultJob"
        Client = "bareos-fd"
        Client Run Before Job = /bin/false
    }

> Now you can add the reschedule directives as follows
> 
> .. code-block:: sh
>    :caption: Job with reschedule directives
> 
>     Job {
>         Name = "backup-bareos-fd"
>         JobDefs = "DefaultJob"
>         Client = "bareos-fd"
>         Client Run Before Job = /bin/false
>         Reschedule On Error = yes
>         Reschedule Times = 4
>         Reschedule Interval = 5 seconds
>     }

Reviewer: ```suggestion
```
